GVFS/GVFS/CommandLine/UpgradeVerb.cs

```diff
@@ -0,0 +1,309 @@
+using CommandLine;
+using GVFS.Common;
+using GVFS.Common.Git;
+using GVFS.Common.Tracing;
+using GVFS.Upgrader;
+using System;
+using System.Diagnostics;
+using System.IO;
+
+namespace GVFS.CommandLine
+{
+    [Verb(UpgradeVerbName, HelpText = "Checks if a new GVFS release is available.")]
+    public class UpgradeVerb : GVFSVerb
+    {
+        private const string UpgradeVerbName = "upgrade";
+        private ITracer tracer;
+        private ProductUpgrader upgrader;
+        private InstallerPreRunChecker prerunChecker;
+        private ProcessLauncher processWrapper;
+
+        public UpgradeVerb(
+            ProductUpgrader upgrader,
+            ITracer tracer,
+            InstallerPreRunChecker prerunChecker,
+            ProcessLauncher processWrapper,
+            TextWriter output)
+        {
+            this.upgrader = upgrader;
+            this.tracer = tracer;
+            this.prerunChecker = prerunChecker;
+            this.processWrapper = processWrapper;
+            this.Output = output;
+        }
+
+        public UpgradeVerb()
+        {
+            JsonTracer jsonTracer = new JsonTracer(GVFSConstants.GVFSEtwProviderName, "UpgradeVerb");
+            string logFilePath = GVFSEnlistment.GetNewGVFSLogFileName(
+                ProductUpgrader.GetLogDirectoryPath(),
+                GVFSConstants.LogFileTypes.UpgradeVerb);
+            jsonTracer.AddLogFileEventListener(logFilePath, EventLevel.Informational, Keywords.Any);
+
+            this.tracer = jsonTracer;
+            this.prerunChecker = new InstallerPreRunChecker(this.tracer);
+            this.processWrapper = new ProcessLauncher();
+            this.Output = Console.Out;
+            this.upgrader = new ProductUpgrader(ProcessHelper.GetCurrentProcessVersion(), this.tracer);
+        }
+
+        [Option(
+            "confirm",
+            Default = false,
+            Required = false,
+            HelpText = "Pass in this flag to actually install the newest release")]
+        public bool Confirmed { get; set; }
+
+        public override string EnlistmentRootPathParameter { get; set; }
+
+        protected override string VerbName
+        {
+            get { return UpgradeVerbName; }
+        }
+        
+        public override void Execute()
+        {
+            ReturnCode exitCode = ReturnCode.Success;
+            if (!this.TryRunProductUpgrade())
+            {
+                exitCode = ReturnCode.GenericError;
+                this.ReportErrorAndExit(this.tracer, exitCode, string.Empty);
+            }
+        }
+        
+        private bool TryRunProductUpgrade()

```

Comment:
Under what conditions should `TryRunProductUpgrade` return true?

I noticed that currently there are some scenarios where `TryRunProductUpgrade` will return true even if the product upgrade was not run.

Reply:
Returns true in the following cases.

1. during `gvfs upgrade` 
When upgrade ring is set to “None”
No error checking for upgrade and no new updates are available.
No error checking for upgrade and new updates are available.

2. during `gvfs upgrade --confirm`
When upgrade ring is set to “None”
No error checking for upgrade and no new updates are available.
Upgrade is installable and there were no errors running the installer